core/src/main/scala/org/bitcoins/core/protocol/script/ControlBlock.scala

```diff
@@ -31,9 +27,60 @@ case class ControlBlock(bytes: ByteVector) extends NetworkElement {
   }
 }
 
+case class TapscriptControlBlock(bytes: ByteVector) extends ControlBlock {
+  require(TapscriptControlBlock.isValid(bytes),
+          s"Invalid leaf version for tapscript control block, got=$bytes")
+}
+
+/** A control block that does not have a leaf version defined as per BIP342
+  * This is needed for future soft fork compatability where we introduce new leaf versions
+  * to correspond to new spending rules
+  */
+case class UnknownControlBlock(bytes: ByteVector) extends ControlBlock
+
 object ControlBlock extends Factory[ControlBlock] {
 
   override def fromBytes(bytes: ByteVector): ControlBlock = {
-    new ControlBlock(bytes)
+    TapscriptControlBlock(bytes)
```

Comment:
This should cover both cases, known and unknown

```scala
    if (TapscriptControlBlock.isValid(bytes)) {
      TapscriptControlBlock(bytes)
    } else {
      UnknownControlBlock(bytes)
    }
```

Reply:
We don't want to do this because there are certain cases where a `P2WSHWitnessV0` is a valid control block. I've been battling with this on #3769, unfortunately some witnesses can have the correct length (`33 + 32m` bytes in length).

I think the only way we can properly support this is if we _know_ we are spending a `TaprootScriptPubKey`, else we will get false positives on `UnknownControlBlock`

Reply:
Yeah maybe a parent type could be made that encapsulates `P2WSHWitnessV0` and `TaprootScriptPath` that we parse to and when evaluating we cast to the correct one based on the spk

Reply:
Idk, we can get it right 99% of the time just not 100% of the time. I think it will be very rare in production that the heurstics are incorrect, we just test a lot of corner cases in our property based tests. I think we leave it for now and revisit when i implement failure cases

Reply:
Is this the root cause of the `lndRpcTest/test` failures?

Reply:
> Is this the root cause of the lndRpcTest/test failures?

I believe so

Reply:
> Is this the root cause of the lndRpcTest/test failures?

I am not too sure, adding the ability to parse these witnesses breaks the tests